File: packages/node/src/builders.js
const NO_DEFAULT = Symbol('NO_DEFAULT');

class TrapBuilder {
    constructor(ts, path) {
        this.ts = ts;
        this.config = {
            path: path,
            methods: ["GET", "POST"],
            intent: null,
            "response.authenticated": JSON.parse(JSON.stringify(this.ts.default_responses["authenticated"])),
            "response.unauthenticated": JSON.parse(JSON.stringify(this.ts.default_responses["unauthenticated"])),
        };
    }

    methods(...args) {
        this.config.methods = args;
        return this;
    }

    intent(intent) {
        this.config.intent = intent;
        return this;
    }

    _respond(key, { status = null, body = null, mime_type = null, template = null } = {}) {
        key = "response." + key;
        if (template) {
            if (status !== null || body !== null || mime_type !== null) {
                throw new Error("response_builder: `template` cannot be used together with `status`, `body`, or `mime_type`.");
            }

            const tmpl = this.ts._templates[template];
            if (!tmpl) {
                throw new Error(`response_builder: template '${template}' not found.`);
            }

            this.config[key] = JSON.parse(JSON.stringify(tmpl));
        } else {
            if (status) {
                this.config[key]["status_code"] = status;
            }

            if (body) {
                this.config[key]["response_body"] = body;
            }

            if (mime_type) {
                this.config[key]["mime_type"] = mime_type;
            }
        }

        return this;
    }

    respond({ status = null, body = null, mime_type = null, template = null } = {}) {
        this._respond("authenticated", { status, body, mime_type, template });
        return this;
    }

    if_unauthenticated({ status = null, body = null, mime_type = null, template = null } = {}) {
        this._respond("unauthenticated", { status, body, mime_type, template });
        return this;
    }

    build() {
        return this.config;
    }
}

class WatchBuilder {
    constructor(path) {
        this.path = path;
        this._query = {};
        this._body = {};
    }

    query(name, { defaultValue = NO_DEFAULT, intent = null } = {}) {
        this._query[name] = { default: defaultValue, intent: intent };
        return this;
    }

    body(name, { defaultValue = NO_DEFAULT, intent = null } = {}) {
        this._body[name] = { default: defaultValue, intent: intent };
        return this;
    }

    build() {
        return {
            path: this.path,
            query_fields: this._query,
            body_fields: this._body
        };
    }
}

module.exports = { TrapBuilder, WatchBuilder, NO_DEFAULT };

File: packages/go/builders.go
package trappsec

import "strings"

type ResponseTemplate struct {
	StatusCode int
	Body       any
	MIMEType   string
}

type ResponseConfig struct {
	Status   int
	Body     any
	MIMEType string
	Template string
}

type TrapConfig struct {
	Path                    string
	Methods                 []string
	Intent                  string
	ResponseAuthenticated   ResponseTemplate
	ResponseUnauthenticated ResponseTemplate
}

type WatchFieldRule struct {
	Default any
	Intent  string
}

type WatchConfig struct {
	Path        string
	QueryFields map[string]WatchFieldRule
	BodyFields  map[string]WatchFieldRule
}

type FoundField struct {
	Type   string `json:"type"`
	Field  string `json:"field"`
	Value  any    `json:"value"`
	Intent string `json:"intent,omitempty"`
}

type TrapBuilder struct {
	ts     *Sentry
	config TrapConfig
}

func newTrapBuilder(ts *Sentry, path string) *TrapBuilder {
	return &TrapBuilder{
		ts: ts,
		config: TrapConfig{
			Path:                    path,
			Methods:                 []string{"GET", "POST"},
			Intent:                  "",
			ResponseAuthenticated:   cloneTemplate(ts.defaultResponses["authenticated"]),
			ResponseUnauthenticated: cloneTemplate(ts.defaultResponses["unauthenticated"]),
		},
	}
}

func (b *TrapBuilder) Methods(methods ...string) *TrapBuilder {
	cleaned := make([]string, 0, len(methods))
	for _, m := range methods {
		m = strings.TrimSpace(strings.ToUpper(m))
		if m != "" {
			cleaned = append(cleaned, m)
		}
	}
	if len(cleaned) > 0 {
		b.config.Methods = cleaned
		b.ts.touchConfig()
	}
	return b
}

func (b *TrapBuilder) Intent(intent string) *TrapBuilder {
	b.config.Intent = intent
	b.ts.touchConfig()
	return b
}

func (b *TrapBuilder) Respond(cfg ResponseConfig) *TrapBuilder {
	b.applyResponse("authenticated", cfg)
	return b
}

func (b *TrapBuilder) IfUnauthenticated(cfg ResponseConfig) *TrapBuilder {
	b.applyResponse("unauthenticated", cfg)
	return b
}

func (b *TrapBuilder) Build() TrapConfig {
	return b.config
}

func (b *TrapBuilder) applyResponse(key string, cfg ResponseConfig) {
	var target *ResponseTemplate
	if key == "authenticated" {
		target = &b.config.ResponseAuthenticated
	} else {
		target = &b.config.ResponseUnauthenticated
	}

	if cfg.Template != "" {
		b.ts.mu.RLock()
		t, ok := b.ts.templates[cfg.Template]
		b.ts.mu.RUnlock()
		if !ok {
			panic("response_builder: template not found: " + cfg.Template)
		}
		*target = cloneTemplate(t)
		b.ts.touchConfig()
		return
	}

	if cfg.Status > 0 {
		target.StatusCode = cfg.Status
	}
	if cfg.Body != nil {
		target.Body = cfg.Body
	}
	if cfg.MIMEType != "" {
		target.MIMEType = cfg.MIMEType
	}
	b.ts.touchConfig()
}

type WatchBuilder struct {
	ts     *Sentry
	config WatchConfig
}

func newWatchBuilder(ts *Sentry, path string) *WatchBuilder {
	return &WatchBuilder{
		ts: ts,
		config: WatchConfig{
			Path:        path,
			QueryFields: map[string]WatchFieldRule{},
			BodyFields:  map[string]WatchFieldRule{},
		},
	}
}

func (b *WatchBuilder) Query(name string, defaultValue any, intent string) *WatchBuilder {
	b.config.QueryFields[name] = WatchFieldRule{Default: defaultValue, Intent: intent}
	b.ts.touchConfig()
	return b
}

func (b *WatchBuilder) Body(name string, defaultValue any, intent string) *WatchBuilder {
	b.config.BodyFields[name] = WatchFieldRule{Default: defaultValue, Intent: intent}
	b.ts.touchConfig()
	return b
}

func (b *WatchBuilder) Build() WatchConfig {
	return b.config
}
